### PR TITLE
 Add: flatten(), flatten one level of nesting

### PR DESCRIPTION
**Pull request by Kijewski**
Inspired by [more_itertools.flatten()][1].

This PR depends on #16.

[1]: https://more-itertools.readthedocs.io/en/stable/api.html#more_itertools.flatten

### CODE REVIEW

File: include/rx/ranges.hpp [hunk 1]
```diff
@@ -735,7 +735,8 @@ seq(T, U)->seq<T, U>;
 template <class T>
 struct repeat {
     T value;
-    constexpr explicit repeat(T value) noexcept : value(std::move(value)) {}
+    template <class V>
+    constexpr explicit repeat(V&& value) noexcept : value(std::forward<V>(value)) {}
```
Comment (Hazurl) [inline include/rx/ranges.hpp]: This prevents the copy/move constructor to be used.

File: include/rx/ranges.hpp [hunk 2]
```diff
@@ -1998,6 +1972,114 @@ struct reverse {
...
+
+template <>
+struct flatten <1> {
+    template <class R>
+    struct Range {
+        using S = get_range_type_t<get_output_type_of_t<R>>;
+        using output_type = get_output_type_of_t<S>;
+
+        static constexpr bool is_finite = is_finite_v<R> && is_finite_v<S>;
+        static constexpr bool is_idempotent = is_idempotent_v<S>;
```
Comment (Hazurl) [inline include/rx/ranges.hpp:2385]: not `is_idempotent_v<R> && is_idempotent_v<S>` ?

File: include/rx/ranges.hpp [hunk 3]
```diff
@@ -1998,6 +1972,114 @@ struct reverse {
...
+
+        R outer_range;
+        RX_OPTIONAL<S> inner_range;
+
+        template <class Rx>
+        constexpr explicit Range(Rx input) : outer_range(std::forward<Rx>(input)) {
```
Comment (Hazurl) [inline include/rx/ranges.hpp]: forgot `&&` in the parameter : `Range(Rx&& input)`

File: include/rx/ranges.hpp [hunk 4]
```diff
@@ -1998,6 +1972,114 @@ struct reverse {
...
+            }
+
+            auto r = outer_range.size_hint();
+            auto s = inner_range->size_hint();
+            auto rs = r * s;
+            if (!r) {
+                return s;
+            } else if (!s) {
+                return r;
+            } else if (rs < r || rs < s) {
```
Comment (Hazurl) [inline include/rx/ranges.hpp]: `r == std::numeric_limits<size_t>::max() || s == std::numeric_limits<size_t>::max()` would be clearer IMO, you could also move the calcul of `rs` in the final else branch.

File: test/test_ranges.cpp [hunk 5]
```diff
@@ -233,6 +233,10 @@ TEST_CASE("ranges fill") {
     std::string b;
     fill('b') | first_n(5) | append(b);
     CHECK(b == "bbbbb");
+
+    int v = 7;
+    CHECK((fill(v) | take(5) | sum()) == 7*5);
```
Comment (Hazurl) [inline test/test_ranges.cpp]: `... == v*5` or else the use of `v` is kinda useless

**Comment by Hazurl** [comment]
Also a duplicate of `repeat_n`'s removal from #16

File: include/rx/ranges.hpp (hunk 4 above)
Comment (simonask) [reply include/rx/ranges.hpp]: You can use `R::is_finite` and `S::is_finite` with `if constexpr` to statically return `numeric_limits<size_t>::max()`.

Note that the contract for `size_hint()` was changed in #10 to return `numeric_limits<size_t>::max()` when the output length is unknown. This was done to keep the possibility of sometimes treating the hint as an exact value when the output length is actually precisely known, which in turn would allow specializations of combinators that can be more easily auto-vectorized.

File: include/rx/ranges.hpp [hunk 6]
```diff
@@ -763,43 +764,6 @@ template <class T>
...
-    }
-};
-template <class T>
-repeat_n(size_t, T &&)->repeat_n<T>;
-
-template <class T>
-[[nodiscard]] constexpr auto fill_n(size_t n, T&& v) noexcept {
-    return repeat_n(n, std::forward<T>(v));
-}
-
```
Comment (simonask) [inline include/rx/ranges.hpp]: Why was this removed?

File: include/rx/ranges.hpp (hunk 1 above)
Comment (Kijewski) [reply include/rx/ranges.hpp]: I still don't quite get it. Should I add explicitly defaulted (copy|move) (ctors|assignment operators) in here? As I said: IMO one should always add them if copying is a trivial thing. If not either explicitly delete them, or call swap in the move (ctor|assignment operator).

File: include/rx/ranges.hpp (hunk 2 above)
Comment (Kijewski) [reply include/rx/ranges.hpp:2385]: R::get() is only called once when needed. S::get() gets called multiple times, as often as the user calls get().

File: test/test_ranges.cpp (hunk 5 above)
Comment (Kijewski) [reply test/test_ranges.cpp]: v is entirely useless in here, but it should still work.

File: include/rx/ranges.hpp (hunk 6 above)
Comment (Kijewski) [reply include/rx/ranges.hpp]: It was moved further to the bottom, because it can only be defined after struct take: https://github.com/simonask/rx-ranges/pull/17/files#diff-d9c605d1767259c15ff78a379ec1212bR955

Comment (simonask) [reply include/rx/ranges.hpp]: Oh sorry, missed that.

File: include/rx/ranges.hpp (hunk 1 above)
Comment (simonask) [reply include/rx/ranges.hpp]: It's probably easier to just take `V` by value and have an explicitly defaulted default constructor. It's fine if people have to manually `std::move` things in.

File: test/test_ranges.cpp (hunk 5 above)
Comment (simonask) [reply test/test_ranges.cpp]: Why not use it in the test for clarity, though? :)

**Comment by simonask** [comment]
@Kijewski If you can address remaining comments and merge `version-2` into this branch, I would love to get this in! :-)

**Comment by Kijewski** [comment]
I'll address the comments after #16 is merged. Working in two branches that share some commits is a little irritating. :)

**Comment by Kijewski** [comment]
Rebased.

**Review by simonask — APPROVED** [review]
Awesome 👌
